pkg/forwarder/domain_forwarder.go

```diff
@@ -79,10 +81,22 @@ func (f *domainForwarder) retryTransactions(retryBefore time.Time) {
 	droppedRetryQueueFull := 0
 	droppedWorkerBusy := 0
 
-	sort.Sort(byCreatedTimeAndPriority(f.retryQueue))
-	totalPayloadsSize := 0
+	var transactions []Transaction
+	var err error
+	if f.optionalTransactionContainer != nil {
+		f.transactionContainerMutex.Lock()
+		transactions, err = f.optionalTransactionContainer.ExtractTransactions()
+		f.transactionContainerMutex.Unlock()
```

Comment:
The mutex is required only because of the method `sendHTTPTransactions` (`retryTransactions` and `requeueTransaction` cannot be called concurrently, see https://github.com/DataDog/datadog-agent/blob/0fc3e11eb24fe5710d440cfccc22b644692fa08d/pkg/forwarder/domain_forwarder.go#L188-L201).

The lock was initially inside `transactionContainer` but I moved the lock outside as it needed because of the implementation of domain_forwarder.

Reply:
My main comment is that since `optionalTransactionContainer` maintain indexes it should protect itself from concurrency (especially in go). I feel it's an issue waiting to happen in future change when we use the class again.

Reply:
I do not think `transactionContainer` should be thread safe by default, but I do not have a strong opinion on that.
Code updated.